### PR TITLE
tracker: Add a file output

### PR DESCRIPTION
**Pull request by HaoZeke**
Also update the README and the gitignore

### CODE REVIEW

File: track/__main__.py [hunk 1]
```diff
@@ -48,31 +47,49 @@ def utc2localtime(utc):
 
 online = None
 last_offline = None
+
+printName = True
 while True:
-    if contact_id in ['me', 'self']:
+    try:
```
Comment (gentoo-root) [inline track/__main__.py:53]: `with` should be used instead of `try-finally` for working with files.

File: track/__main__.py [hunk 2]
```diff
@@ -48,31 +47,49 @@ def utc2localtime(utc):
...
+        # File IO
+        dataFile = open('trackerLogs.txt', 'a+', encoding = 'utf-8')
```
Comment (gentoo-root) [inline track/__main__.py:55]: There is no need to re-open a file at each iteration. It could be opened outside the loop.

File: track/__main__.py [hunk 3]
```diff
@@ -48,31 +47,49 @@ def utc2localtime(utc):
...
         # https://github.com/LonamiWebs/Telethon/issues/1024
-        contact = client.get_me()
-    else:
-        contact = client.get_entity(contact_id)
+            contact = client.get_me()
+        else:
+            contact = client.get_entity(contact_id)
+            user = contact.first_name+' '+contact.last_name
+        if printName == True:
+            print(user, file = dataFile)
```
Comment (gentoo-root) [inline track/__main__.py:64]: If `contact_id` was `'me'`, `user` is unset here, leading to a crash.

File: track/__main__.py [hunk 4]
```diff
@@ -48,31 +47,49 @@ def utc2localtime(utc):
...
+                print(f'~{datetime.now().strftime(DATETIME_FORMAT)}: User went online.')
+        else:
+            if online != False:
+                online = False
+                print(f'~{datetime.now().strftime(DATETIME_FORMAT)}: User went offline.', file = dataFile)
+                print(f'~{datetime.now().strftime(DATETIME_FORMAT)}: User went offline.')
+            last_offline = None
+        sleep(15)
+        printName = False
+        dataFile.flush()
```
Comment (gentoo-root) [inline track/__main__.py:93]: Doesn't seem useful to me. The file is closed one line below in the `finally` block.

File: track/__main__.py [hunk 5]
```diff
@@ -48,31 +47,49 @@ def utc2localtime(utc):
...
-            print(f'~{datetime.now().strftime(DATETIME_FORMAT)}: User went online.')
-    else:
-        if online != False:
-            online = False
-            print(f'~{datetime.now().strftime(DATETIME_FORMAT)}: User went offline.')
-        last_offline = None
-    sleep(15)
+            elif last_offline != contact.status.was_online:
+                if last_offline is not None:
+                    print(f'={utc2localtime(contact.status.was_online).strftime(DATETIME_FORMAT)}: User went offline after being online for short time.', file = dataFile)
```
Comment (gentoo-root) [inline track/__main__.py:74]: Duplicate code is all around the file after your changes. Every print is copied and pasted. This is not the coding style I can accept.

File: track/__main__.py [hunk 6]
```diff
@@ -48,31 +47,49 @@ def utc2localtime(utc):
...
-            online = False
-            print(f'={utc2localtime(contact.status.was_online).strftime(DATETIME_FORMAT)}: User went offline.')
-        elif last_offline != contact.status.was_online:
-            if last_offline is not None:
-                print(f'={utc2localtime(contact.status.was_online).strftime(DATETIME_FORMAT)}: User went offline after being online for short time.')
-            else:
+        if isinstance(contact.status, UserStatusOffline):
+            if online != False:
+                online = False
+                print(f'={utc2localtime(contact.status.was_online).strftime(DATETIME_FORMAT)}: User went offline.', file = dataFile)
```
Comment (gentoo-root) [inline track/__main__.py:70]: No spaces should be around `=` that is used to indicate a keyword argument.

https://www.python.org/dev/peps/pep-0008/#other-recommendations

**Review by gentoo-root — CHANGES_REQUESTED** [review]
I'm sorry, but I can't accept this pull request. Not only the code quality is not acceptable for this project, but actually this script doesn't need this feature.

If you want the output to a file and to stdout, you can always run the script as follows:

```
# Output both to the stdout and the file, appending if it exists:
python3 -m track me | tee -a tracker-me.log
```

Without any code modification, you have even more flexibility. You can freely choose any filename you want. On the other hand, your implementation has a hardcoded filename, and it's impossible to run multiple instances of the tracker: either the file will contain a mix of all outputs, or you have to `cd` to different directories before running. It breaks the basic operation of the script. Please use `tee` as suggested above, it's more flexible and doesn't require any change in code.

Second, printing the user name to the output file breaks the format. Especially if you restart the script, and the user name appears on some random lines in the middle of the file. It will be impossible to parse it. If you really need it, consider writing a separate tool, also based on telethon, that would just fetch the user name and print it to stdout. You will be able to run it before this script. This feature is outside the scope and use case of this script.

Thanks for your effort, I hope that you'll solve your problem using the information above. I also left a bunch of comments on your code — you may find them useful to write better code. Feel free to ask questions and good luck!

File: track/__main__.py (hunk 4 above)
Comment (HaoZeke) [reply track/__main__.py:93]: This is actually the only change I can really justify, without it there is no way to check the `.txt` for current changes without stopping the script, say for when you need to check from another PC, without using `tmux`

**Comment by HaoZeke** [comment]
Ah sorry, I was actually just sort of very trigger happy with `git push`. I am abandoning possible updates to this since it feels slightly stalkerish and brought on by hospital paranoia (hospitalized for a stupid accident).

Thanks for your very detailed review and kind comments!

**Pull request closed without merging**